File: src/main/java/com/urbanfix/service/InterFaces/ComplaintService.java
package com.urbanfix.service.InterFaces;

import com.urbanfix.dto.CategoryAnalyticsResponse;

//import org.hibernate.query.Page;

import com.urbanfix.dto.ComplaintResponse;
import com.urbanfix.dto.CreateComplaintRequest;
import com.urbanfix.dto.DashboardStatsResponse;
import com.urbanfix.dto.MonthlyComplaintAnalyticsResponse;
import com.urbanfix.dto.UpdateComplaintRequest;
import com.urbanfix.dto.UpdateComplaintStatusRequest;
import com.urbanfix.enums.ComplaintStatus;

import org.springframework.data.domain.Page;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface ComplaintService 
    {

        // Create a new complaint
        ComplaintResponse createComplaint(CreateComplaintRequest request,MultipartFile image);

        //List<ComplaintResponse> getAllComplaints();
        ComplaintResponse getComplaintById(Long complaintId);
        ComplaintResponse updateComplaint( Long complaintId, UpdateComplaintRequest request);
        void deleteComplaint(Long complaintId);
        // Returns the authenticated user's complaints with optional filters.
        Page<ComplaintResponse> getMyComplaints(
                        int page,
                        int size,
                        String sortBy,
                        String direction,
                        ComplaintStatus status,
                        String category,
                        String keyword);

        Page<ComplaintResponse> getAllComplaints(int page ,int size, String sortBy, String direction,
                                                ComplaintStatus status,String keyword);
        
        Page<ComplaintResponse> searchComplaints(String keyword,int page,int size);

        // Update the status of an existing complaint
        ComplaintResponse updateComplaintStatus(Long complaintId,UpdateComplaintStatusRequest request);

        // Retrieve all complaints for the admin dashboard
        Page<ComplaintResponse> getAllComplaintsForAdmin(
                ComplaintStatus status,
                String  category,
                int page,
                int size,
                String sortBy,
                String sortDirection);
        
        // Retrieve complaint statistics for the admin dashboard
        DashboardStatsResponse getDashboardStatistics();

                /**
         * Retrieves complaint count grouped by category.
         */
        List<CategoryAnalyticsResponse> getCategoryAnalytics();
        List<MonthlyComplaintAnalyticsResponse> getMonthlyComplaintAnalytics();
    
    }

File: urbanfix-frontend/src/services/complaintService.js
import api from './api';

export async function createComplaint(complaint, image) {
  const formData = new FormData();

  Object.entries(complaint).forEach(([key, value]) => {
    formData.append(key, value);
  });

  if (image) {
    formData.append('image', image);
  }

  const { data } = await api.post('/complaints', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return data;
}

export async function getComplaints(params) {
  const { data } = await api.get('/complaints', { params });
  return data;
}

export async function getMyComplaints(params) {
  const { data } = await api.get('/complaints/my', { params });
  return data;
}

export async function getComplaintById(complaintId) {
  const { data } = await api.get(`/complaints/${complaintId}`);
  return data;
}

export async function updateComplaint(complaintId, complaint) {
  const { data } = await api.put(`/complaints/${complaintId}`, complaint);
  return data;
}

export async function deleteComplaint(complaintId) {
  const { data } = await api.delete(`/complaints/${complaintId}`);
  return data;
}

export async function searchComplaints(keyword, params) {
  const { data } = await api.get('/complaints/search', { params: { keyword, ...params } });
  return data;
}

// The backend authorizes this operation for administrators only.
export async function updateComplaintStatus(complaintId, status) {
  const { data } = await api.patch(`/complaints/${complaintId}/status`, { status });
  return data;
}
